packages/realm/src/schema/normalize.ts

```diff
@@ -353,42 +354,52 @@ function optionalIsImplicitlyTrue(type: string, objectType: string | undefined):
 /**
  * Determine whether a property always is implicitly non-optional (non-nullable).
  */
-function optionalIsImplicitlyFalse(type: string, objectType: string | undefined): boolean {
+function isNeverOptional(type: string, objectType: string | undefined): boolean {
   return (type === "list" || type === "set" || type === "linkingObjects") && isUserDefined(objectType);
 }
 
 /**
  * Determine whether a string ends with a shorthand collection ('[]' or '{}' or '<>').
  */
 function hasCollectionSuffix(input: string): boolean {
-  const end = input.substring(input.length - 2);
-  return !!COLLECTION_SHORTHAND_TO_NAME[end];
+  const suffix = input.substring(input.length - COLLECTION_SUFFIX_LENGTH);
+  return !!COLLECTION_SHORTHAND_TO_NAME[suffix];
 }
 
 /**
  * Assert that shorthand notation is not being used.
  */
 function assertNotUsingShorthand(input: string | undefined, info: PropertyInfo): void {
-  const shorthands: string[] = [];
-
-  if (input && hasCollectionSuffix(input)) {
-    shorthands.push(input.substring(input.length - 2));
-    input = input.substring(0, input.length - 2);
-  }
-
-  if (input?.endsWith("?")) {
-    shorthands.push("?");
+  if (!input) {
+    return;
   }
 
+  const shorthands = extractShorthands(input);
   assert(
-    !shorthands.length,
+    shorthands.length === 0,
     propError(
       info,
       `Cannot use shorthand '${shorthands.join("' and '")}' in 'type' or 'objectType' when defining property objects.`,
     ),
   );
 }
 
+/**
+ * Extract the shorthand markers used in the input.
+ */
+function extractShorthands(input: string): string[] {
+  const shorthands: string[] = [];
+  if (hasCollectionSuffix(input)) {
+    shorthands.push(input.substring(input.length - COLLECTION_SUFFIX_LENGTH));
+    input = input.substring(0, input.length - COLLECTION_SUFFIX_LENGTH);
+  }
+  if (input.endsWith("?")) {
+    shorthands.push("?");
+  }
+
+  return shorthands;
+}
```

Comment:
Great candidate for a separate function 👍

Reply:
Indeed, thanks 🙂